lib/mongoid/contextual/mongo.rb

```diff
@@ -465,23 +465,28 @@ def take!
       #   # => { [ 1, 2 ] => 1 }
       #
       # @param [ String | Symbol ] field The field name.
+      # @param [ Boolean ] :splat_arrays Whether to tally array
+      #   member values individually. Default false.
       #
       # @return [ Hash ] The hash of counts.
-      def tally(field)
+      def tally(field, splat_arrays: false)
         name = klass.cleanse_localized_field_names(field)
+        is_translation = "#{name}_translations" == field.to_s
+
+        # Must add a $project stage when using $unwind with nested fields
+        projected = 'p' if splat_arrays && (is_translation || name.include?('.'))
 
         fld = klass.traverse_association_tree(name)
-        pipeline = [ { "$group" => { _id: "$#{name}", counts: { "$sum": 1 } } } ]
-        pipeline.unshift("$match" => view.filter) unless view.filter.blank?
+        pipeline = []
+        pipeline << { "$match" => view.filter } if view.filter.present?
+        pipeline << { "$project" => { "#{projected}" => "$#{name}" } } if projected
```

Comment:
Having a `$project` stage in the middle of a pipeline can negatively impact performance.

I would be extremely wary of introducing a change like this

Reply:
See https://www.mongodb.com/docs/rapid/core/aggregation-pipeline-optimization/#-project-stage-placement

Reply:
@alexbevi if you don't use `$project`, the `$unwind` on nested fields will yield a null result. I have raised https://jira.mongodb.org/browse/SERVER-73888 for this because I believe it is a bug; other operators such as `$group` handle nested fields just fine.

as you can see in the code, the `$project` is only used in the case it needed:

- the `:splat_arrays` arg is true (note it is false by default) *AND*
- the field is a nested field (either `.` is present, or it is translation)

```ruby
        # Must add a $project stage when using $unwind with nested fields
        projected = 'p' if splat_arrays && (is_translation || name.include?('.'))
```

Note that the code in my PR does *not* apply the `$project` stage under default conditions, so there is no behavior change for any existing usage of this method.

Re: performance, fortunately I have a large production database we can try it and see what happens😄. Using it on a collection with 109 million+ records:

```
# non-nested array<string> field, does *not* use $project
Customer.tally('tags', splat_arrays: true)
#=> 481 seconds
#=> 761774 tally keys, total tallied value count 8031404

# nested field (array<document> -> string), uses $project
Customer.tally('addresses.city', splat_arrays: true)
#=> 592 seconds
#=> 65362 tally keys, total tallied value count 3284490
```

CPU, memory, disk util etc. were all equivalent & healthy on both throughout. Not apples-to-apples obviously, but clearly, nothing catastrophic happens when adding `$project` where it is needed due.
![image](https://user-images.githubusercontent.com/27655/218155812-566bfb02-e848-4765-9eb4-036359e8b54c.png)

TLDR;
![image](https://user-images.githubusercontent.com/27655/218152511-d50579bf-02b7-41ab-bf20-e1459a5e1741.png)

Reply:
One final note, the link you posted: https://www.mongodb.com/docs/rapid/core/aggregation-pipeline-optimization/#-project-stage-placement does not say putting a `$project` in the beginning will *hurt* performance, it says it will *not improve* performance. As per above, the reason we are using `$project` is nothing related to performance, it is required as per [SERVER-73888](https://jira.mongodb.org/browse/SERVER-73888).